incubator/hnc/pkg/controllers/object_controller.go

```diff
@@ -283,53 +265,81 @@ func (r *ObjectReconciler) onDelete(ctx context.Context, log logr.Logger, nnm ty
 	// so no need to do this recursively. TODO: maybe do a search on the label value instead?
 	// Parallelize?
 	for _, child := range r.getChildNamespaces(nnm.Namespace) {
-		// Try to find the copied objects, if they exist
-		copiedNnm := types.NamespacedName{Namespace: child, Name: nnm.Name}
-		copied := &unstructured.Unstructured{}
-		copied.SetGroupVersionKind(r.GVK)
-		err := r.Get(ctx, copiedNnm, copied)
+		// Try to find the propagated objects, if they exist
+		propagatedNnm := types.NamespacedName{Namespace: child, Name: nnm.Name}
+		propagated := &unstructured.Unstructured{}
+		propagated.SetGroupVersionKind(r.GVK)
+		err := r.Get(ctx, propagatedNnm, propagated)
 
 		// Is it already gone?
 		if errors.IsNotFound(err) {
 			continue
 		}
 		// Some other error?
 		if err != nil {
-			log.Error(err, "Couldn't read copied object that needs to be deleted", "name", copiedNnm)
+			log.Error(err, "Couldn't read propagated object that needs to be deleted", "name", propagatedNnm)
 			return err
 		}
 
 		// TODO: double-check the label - or maybe just call deleteObsolete?
 
 		// Delete the copy
-		log.Info("Deleting", "propagated", copiedNnm)
-		if err := r.Delete(ctx, copied); err != nil {
-			log.Error(err, "Coudln't delete", "copy", copied)
+		log.Info("Deleting", "propagated", propagatedNnm)
+		if err := r.Delete(ctx, propagated); err != nil {
+			log.Error(err, "Coudln't delete", "propagated", propagated)
 			return err
 		}
 	}
 
 	return nil
 }
 
-func copyObject(inst *unstructured.Unstructured) *unstructured.Unstructured {
-	copied := inst.DeepCopy()
-
-	// Clear all irrelevant fields. cf https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.15/#objectmeta-v1-meta
-	copied.SetCreationTimestamp(metav1.Time{})
-	copied.SetDeletionGracePeriodSeconds(nil)
-	copied.SetDeletionTimestamp(nil)
-	copied.SetFinalizers(nil) // TODO: double-check this is the right thing to do?
-	copied.SetGenerateName("")
-	copied.SetGeneration(0)
-	copied.SetInitializers(nil) // TODO: is this correct?
+// canonical returns a canonicalized version of the object - that is, one that has the same name and
+// spec, but with automatically generated fields, the Status, and any HNC-added labels or
+// annotations cleared. The resulting object is suitable to be copied into a new namespace, or two
+// canonicalized objects are suitable for being compared via reflect.DeepEqual.
+//
+// As a side effect, the label and annotation maps are always initialized in the returned value.
+func canonical(inst *unstructured.Unstructured) *unstructured.Unstructured {
```

Comment:
This seems like a lot of things to keep track of throughout k8s releases. Could we possibly just shove the canoical state that was applied last into an annotation similar to how `kubectl apply` works with the `kubectl.kubernetes.io/last-applied-configuration` annotation? My main concern tracking/locking HNC versions based on k8s version. Although, the concern is probably overstated as this likely wont change that often.

Reply:
I'm a bit irrationally weirded out by having objects contain copies of themselves 😐 More practically, some objects could have _two_ copies of themselves - one from `kubectl apply` and one from HNC. Also, it's hard to tell when we should update it...

How about this instead: basically, we only want to keep the spec, name, and non-HNC labels and annotations. How about we document that this is the formal behaviour and _all_ other ObjectMetadata fields will be cleared? Then instead of a brittle blacklist, we have a much smaller and robust whitelist?

Reply:
That works for me. If you'd rather take that on later in a separate issue lmk and i'll lgtm this.

Reply:
Fixed